### ISSUE TITLE
rewrite Curv to use WebGPU

### ISSUE THREAD

**Issue by doug-moen** [curv3d/curv#146]
The proposal is to rewrite Curv to use WebGPU, replacing the use of OpenGL. Since Curv is written in C++, we will use Google's [Dawn](https://dawn.googlesource.com/dawn) library, which has a C++ interface. The Curv shader compiler generates GLSL code, so we will use a library to translate GLSL into SPIR-V, which Dawn accepts as input.

The only Curv code that changes is located in `libcurv/viewer`. This code runs the graphical "viewer" window. If an existing "shader sandbox" app can be found that is written in WebGPU, then the structure of that code can be copied and reused for the Curv viewing window. The viewer code depends on the imgui library for GUI widgets, but I see that there is already a WebGPU port for imgui.

**Rationale:** Currently Curv is stuck at version 3.3 of OpenGL due to MacOS compatibility requirements. There's no way to use advanced features like compute shaders in OpenGL due to MacOS limitations. WebGPU can be thought of as the modern replacement for OpenGL. It is faster and more powerful than OpenGL. It is much easier to use than Vulkan. WebGPU is more portable than Vulkan, since it runs on all desktop and mobile platforms, plus it supports web assembly (WASM) and web browsers.

Once Curv is ported to WebGPU, we can build a faster and more powerful rendering engine based on the use of compute shaders.

**Comment by pbsds** [comment]
An alternative backend is the WGPU C++ api: https://github.com/gfx-rs/wgpu-native
I haven't tried it, but I believe it can run on directx, vulcan, opengl and metal, possibly featuring support for multiple shader languages.

**Comment by doug-moen** [comment]
I agree.There is a C/C++ implementation from Google ([Dawn](https://dawn.googlesource.com/dawn)) that doesn't introduce a Rust dependency. Nice to know that gfx-rs now has a C++ wrapper. WGPU is on track to be finalized some time this year. (Right now it is still an unstable alpha, with breaking changes still planned.) WGPU will open up a richer API, including compute shaders, which will be very useful to Curv.
